packages/block-directory/src/store/controls.js

```diff
@@ -0,0 +1,149 @@
+	return {
+		type: 'SELECT',
+		storeName,
+		selectorName,
+		args,
+	};
+}
+
+/**
+ * Calls a dispatcher using the current state.
+ *
+ * @param {string} storeName      Store name.
+ * @param {string} dispatcherName Dispatcher name.
+ * @param  {Array} args           Selector arguments.
+ *
+ * @return {Object} control descriptor.
+ */
+export function dispatch( storeName, dispatcherName, ...args ) {
+	return {
+		type: 'DISPATCH',
+		storeName,
+		dispatcherName,
+		args,
+	};
+}
+
+/**
+ * Trigger an API Fetch request.
+ *
+ * @param {Object} request API Fetch Request Object.
+ * @return {Object} control descriptor.
+ */
+export function apiFetch( request ) {
+	return {
+		type: 'API_FETCH',
+		request,
+	};
+}
+
+/**
+ * Loads JavaScript
+ * @param {Array} asset the url for the JavaScript.
+ * @param {Function} onLoad callback function on success.
+ * @param {Function} onError callback funciton on failure.
+ */
+const loadScript = ( asset, onLoad, onError ) => {
+	if ( ! asset ) {
+		return;
+	}
+	const existing = document.querySelector( `script[src="${ asset.src }"]` );
+	if ( existing ) {
+		existing.parentNode.removeChild( existing );
+	}
+	const script = document.createElement( 'script' );
+	script.src = typeof asset === 'string' ? asset : asset.src;
+	script.onload = onLoad;
+	script.onerror = onError;
+	document.body.appendChild( script );
+};
+
+/**
+ * Loads CSS file.
+ * @param {*} asset the url for the CSS file.
+ */
+const loadStyle = ( asset ) => {
+	if ( ! asset ) {
+		return;
+	}
+	const link = document.createElement( 'link' );
+	link.rel = 'stylesheet';
+	link.href = typeof asset === 'string' ? asset : asset.src;
+	document.body.appendChild( link );
+};
+
+/**
+ * Load the asset files for a block
+ * @param {Array} assets A collection of URL for the assets
+ * @param {Function} onLoad callback function on success.
+ * @param {Function} onError callback funciton on failure.
+ * @return {Object} control descriptor.
+ */
+export function* loadAssets( assets, onLoad, onError ) {
+	return {
+		type: 'LOAD_ASSETS',
+		assets,
+		onLoad,
+		onError,
+	};
+}
+
+const controls = {
+	SELECT: createRegistryControl( ( registry ) => ( { storeName, selectorName, args } ) => {
+		return registry.select( storeName )[ selectorName ]( ...args );
+	} ),
+	DISPATCH: createRegistryControl( ( registry ) => ( { storeName, dispatcherName, args } ) => {
+		return registry.dispatch( storeName )[ dispatcherName ]( ...args );
+	} ),
+	API_FETCH( { request } ) {
+		return wpApiFetch( { ... request } );
+	},
+	LOAD_ASSETS( { assets } ) {
+		return new Promise( ( resolve, reject ) => {
+			if ( typeof assets === 'object' && assets.constructor === Array ) {
+				let scriptsCount = 0;
+				forEach( assets, ( asset ) => {
+					if ( asset.match( /\.js$/ ) !== null ) {
+						scriptsCount++;
+						loadScript( asset, () => {
+							scriptsCount--;
+							if ( scriptsCount === 0 ) {
+								return resolve( scriptsCount );
```

Comment:
@talldan Although technically feasible, could loading them asynchronously cause JS errors for plugins that require a specific load order? Since the `onload` event doesn't get called until the script is executed, there may be missing object on the window. Thoughts?

Reply:
@StevenDufresne Sorry for the delay, only just spotted this notification.

That's a good point, though my interpretation of the above code is that scripts are loaded concurrently anyway (`loadScript` appends the script tags and is just called from a standard JS loop, there's no waiting for the previous request to finish). My comment was more about refactoring the code to avoid incrementing/decrementing `scriptsCount`.

Reply:
Fair enough, we also need to consider another, more reliable way to determine what constitutes a successful load.  There are some plugins that successfully loads, by this definition, but in reality an error was thrown in the plugin's JS itself rendering the block useless. The user is not aware that the plugin is broken and won't work.

Reply:
True. I wonder if listening for an error event might work:
https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror

Not sure about the browser compatibility, but it does mentioned the URL for the script is part of the error payload.

Another option could be attaching the event listener to the script element itself. Not sure if that'd work.